src/macaron/json_tools.py

```diff
@@ -34,12 +33,22 @@ def json_extract(entry: JsonType, keys: list[str], type_: type[T]) -> T | None:
     target = entry
 
     for index, key in enumerate(keys):
-        if not isinstance(target, dict):
-            logger.debug("Expect the value .%s to be a dict.", ".".join(keys[:index]))
-            return None
-        if key not in target:
-            logger.debug("JSON key '%s' not found in .%s", key, ".".join(keys[:index]))
+        if isinstance(target, dict):
+            if key not in target:
+                logger.debug("JSON key '%s' not found in .%s", key, ".".join(keys[:index]))
+                return None
+        elif isinstance(target, list):
+            if isinstance(key, str):
+                logger.debug("JSON key '%s' of type 'str' cannot extract from list.", key)
+                return None
```

Comment:
I think this check doesn't cover the case where `key` is not a `str` and not `int`. Perhaps it's better to return None if the `key` is not `int`, for example:
```
if not isinstance(key, int):
...
```

Reply:
I agree that in the case of lists, as we always expect `int` keys, it's probably best to check if the key is `int` instead of checking for a different type like `str`.

Reply:
This is no longer applicable after most recent changes.